### PR TITLE
Implemented Refraction

### PR DESCRIPTION
**Pull request by Carlbaum**

### CODE REVIEW

File: src/camera.cc [hunk 1]
```diff
@@ -6,6 +6,12 @@
 #include "point_light.h"
 #include <iostream>
 
+// TODO: Place these somewhere that makes the most sense and remove some?
+const float EPSILON = 0.00001f;
+const float REFRACTION_FACTOR_OI = REFRACTION_INDEX_AIR / REFRACTION_INDEX_GLASS; // outside->in
+const float REFRACTION_FACTOR_IO = REFRACTION_INDEX_GLASS / REFRACTION_INDEX_AIR; // inside->out
+const float CRITICAL_ANGLE = asin(REFRACTION_FACTOR_OI);
```
Comment (novalain) [inline src/camera.cc:13]: yeah probably makes most sense in `camera.h` or `commons.h` right now

File: src/camera.cc [hunk 2]
```diff
@@ -145,6 +156,55 @@ ColorDbl Camera::Shade(Ray& ray, IntersectionPoint& p, Scene& scene) {
...
+  bool is_inside_object = ray.get_refraction_status();
+
+  if ( !is_inside_object ) {
+    Direction T = REFRACTION_FACTOR_OI * I - n*(REFRACTION_FACTOR_OI*I_dot_n +
+        sqrtf(1 - REFRACTION_FACTOR_OI*REFRACTION_FACTOR_OI * (1 - I_dot_n*I_dot_n)));
+    Vertex refraction_point_origin = p.get_position() - n * EPSILON;
+    Ray refraction_ray = Ray(refraction_point_origin,T,ray.get_importance()*0.8f);
+    refraction_ray.set_refraction_status(true);
+    return p.get_material().get_transparence() * Raytrace(refraction_ray, scene);
+  } else { // we are outside of a glass object, trying to go inside
```
Comment (novalain) [inline src/camera.cc:177]: Doesn't make sense with `else` after return statement :)

File: src/camera.cc [hunk 3]
```diff
@@ -145,6 +156,55 @@ ColorDbl Camera::Shade(Ray& ray, IntersectionPoint& p, Scene& scene) {
...
+    // => a = arccos(dot(u,v)), providing u and v are normalized
+    float alpha = acos(I_dot_n);
+
+    if ( alpha > CRITICAL_ANGLE ) { // if total inner reflection
+      Vertex inner_reflection_ray_origin = p.get_position() + n * EPSILON;
+      Direction reflection_direction = I - 2*(glm::dot(I, n))*n;
+      Ray total_inner_reflection_ray = Ray(inner_reflection_ray_origin, reflection_direction, ray.get_importance()*0.8f);
+      total_inner_reflection_ray.set_refraction_status(true);
+      return p.get_material().get_transparence() * Raytrace(total_inner_reflection_ray, scene);
+    } else if ( CRITICAL_ANGLE == alpha ) {
```
Comment (novalain) [inline src/camera.cc:190]: If this should never happen isn't an assert more suitable?

File: src/camera.cc [hunk 4]
```diff
@@ -145,6 +156,55 @@ ColorDbl Camera::Shade(Ray& ray, IntersectionPoint& p, Scene& scene) {
...
+      std::cerr << "THIS SHOULD NEVER (or at least very rarely) BE PRINTED!!!!!" << std::endl;
+      return COLOR_BLACK;
+    }
+    // else exiting glass object!
+    Direction T = REFRACTION_FACTOR_IO * I - n*(-REFRACTION_FACTOR_IO*I_dot_n +
+        sqrtf(1 - REFRACTION_FACTOR_IO*REFRACTION_FACTOR_IO * (1 - I_dot_n*I_dot_n)));
+    // TODO: this can probably be removed, but might be annoying to debug if wrong
+    if ( 1.0f - glm::length(T) > EPSILON ) {
```
Comment (novalain) [inline src/camera.cc:198]: imo `assert(!(1.0f - glm::length(T) > EPSILON));`

File: src/camera.cc [hunk 5]
```diff
@@ -118,10 +128,11 @@ ColorDbl Camera::Shade(Ray& ray, IntersectionPoint& p, Scene& scene) {
     Ray reflection_ray = Ray(reflection_point_origin, reflection_direction);
     // return p.get_material().get_specular() * Raytrace(reflection_ray, scene);
     return p.get_material().get_specular() * Raytrace(reflection_ray, scene);
-  //} else if (p.get_material().get_transparence() > 0.f) {
-  //  return COLOR_PINK;
   }
-
+  // If object has refractive component
+  if (p.get_material().get_transparence() > 0.f) {
```
Comment (novalain) [inline src/camera.cc:133]: If object is both reflective and refractive this won't work?

File: include/ray.h [hunk 6]
```diff
@@ -9,20 +9,28 @@ class Ray {
 private:
   Vertex origin_;
   Direction direction_;
+  float importance_;
```
Comment (novalain) [inline include/ray.h:12]: Why introduce importance sampling before whitted ray tracer works 100%? :)

File: src/camera.cc (hunk 5 above)
Comment (Carlbaum) [reply src/camera.cc:133]: Yeah, but that is a question about refactoring EVERYTHING!

File: src/camera.cc (hunk 2 above)
Comment (Carlbaum) [reply src/camera.cc:177]: You drunk? ;)

File: src/camera.cc (hunk 3 above)
Comment (Carlbaum) [reply src/camera.cc:190]: It can happen

File: include/ray.h (hunk 6 above)
Comment (Carlbaum) [reply include/ray.h:12]: Because in theory, it can bounce around within an object forever.

File: src/camera.cc (hunk 2 above)
Comment (Carlbaum) [reply src/camera.cc:177]: ahh, OK! It's SOO much clearer though

File: include/ray.h (hunk 6 above)
Comment (novalain) [reply include/ray.h:12]: That's why we need an  `int iteration_count;` that's passed and updated recursively and checked against a `const int MAX_ITERATION_DEPTH;` ☕️

File: src/camera.cc (hunk 5 above)
Comment (novalain) [reply src/camera.cc:133]: Hmm, refraction case can maybe handle both reflection and refraction?

File: include/ray.h (hunk 6 above)
Comment (Carlbaum) [reply include/ray.h:12]: Yeah, but it pretty much works the same way now. And we need importance later anyway, depth isn't really necessary if importance drops anyway!

Comment (novalain) [reply include/ray.h:12]: It doesn't drop for pure reflection and still needs to be implemented I think ? https://github.com/novalain/gi-ray/issues/86

File: src/camera.cc (hunk 5 above)
Comment (Carlbaum) [reply src/camera.cc:133]: I think we should have an accumulator to gather the reflective, refractive and diffuse part and treat one at the time. The importance should be split into the three parts according to the specular, transparence and diffuse of the material.

Comment (Carlbaum) [reply src/camera.cc:133]: But yeah, the total inner reflection should probably call a function as it works exactly like the current perfect mirror reflectance. Only difference is that its from within an object.

**Pull request merged**